File: libraries/python/getpatter/services/input_chain.py
"""Inbound audio processing chain for pipeline mode (slice 1 of the
``PipelineStreamHandler`` decomposition — see
``docs/architecture/pipeline-stages.md``).

Owns the stateless-to-STT half of ``on_audio_received``:

    decode (mulaw -> PCM16) -> high-pass/DC-block (opt) -> resample
    -> AEC near-end (opt) -> ``agent.audio_filter``/NS (opt)
    -> AGC (opt) -> VAD frame feed

and returns the processed frame plus at most one VAD event per frame. The
handler keeps everything downstream for this slice (VAD-event handling,
self-hearing gate, inbound ring buffer, ``before_send_to_stt`` hook, STT
feed) so the change stays reviewable.

Stage-order contract (fixed) — canonical WebRTC APM order
``HPF -> AEC -> NS -> VAD -> AGC`` adapted for telephony (AGC pulled ahead
of VAD/STT so the recogniser sees a level-normalised signal):

* The high-pass / DC-block runs FIRST, on the decoded signal BEFORE the
  resampler, so DC and sub-cutoff rumble never reach the echo canceller or
  bias the resampler's interpolation.
* AEC runs before the noise suppressor so the suppressor never disturbs the
  canceller's far-end/near-end alignment.
* ``audio_filter`` (noise suppression) runs AFTER AEC and BEFORE VAD, per the
  :class:`getpatter.providers.base.AudioFilter` docstring ("integrated ...
  before VAD and STT") — the VAD then benefits from the cleaned signal.
* AGC runs AFTER noise suppression and BEFORE VAD/STT: normalising on the
  cleaned signal avoids amplifying suppressed noise, and the level-normalised
  frame helps both the VAD energy gate and the STT recogniser.

Every new stage (HPF, AGC) is OPT-IN, provider-agnostic (runs once,
independent of the chosen STT), and fail-open: a raising stage degrades to
passthrough of its input, logs one WARNING, then DEBUG, and keeps being
attempted — exactly the policy the ``audio_filter`` wrapper uses (a transient
hiccup must not permanently strip processing, a permanent one must never break
the call audio path).

AEC / audio-filter / VAD are resolved through late-bound getter callables
rather than captured at construction: ``PipelineStreamHandler`` populates
``_aec`` / ``_auto_vad`` during ``start()`` (after the chain may already
exist) and the unit suites assign them directly on handler instances.

Mirrors TypeScript ``src/services/input-chain.ts``.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Callable, Optional, TYPE_CHECKING

if TYPE_CHECKING:  # pragma: no cover - typing only
    from getpatter.audio.transcoding import StatefulResampler
    from getpatter.providers.base import AudioFilter, VADEvent, VADProvider

logger = logging.getLogger("getpatter")

# The pipeline's internal processing rate: inbound carrier audio is always
# normalised to PCM16 mono @ 16 kHz before AEC / filter / VAD / STT.
PIPELINE_SAMPLE_RATE: int = 16000


@dataclass(frozen=True)
class InputFrame:
    """Result of pushing one carrier media frame through the input chain.

    Attributes:
        pcm: The decoded / resampled / AEC'd / filtered PCM16 16 kHz bytes —
            exactly what should reach the self-hearing gate and STT.
        vad_event: The VAD event emitted for this frame, if any.
        vad_configured: ``True`` when a VAD provider was consulted for this
            frame — the handler's self-hearing gate only applies then.
    """

    pcm: bytes
    vad_event: "VADEvent | None"
    vad_configured: bool


class InputProcessingChain:
    """HPF -> resample -> AEC -> audio_filter -> AGC -> VAD for inbound audio.

    Args:
        input_is_mulaw_8k: When ``True``, frames are G.711 mu-law @ 8 kHz
            (Twilio always; Telnyx when ``streaming_start`` negotiated PCMU)
            and are decoded + upsampled through a per-call
            :class:`~getpatter.audio.transcoding.StatefulResampler` so the
            ratecv filter state survives chunk boundaries. When ``False``
            the frame is assumed to already be linear PCM16 at
            ``input_sample_rate``.
        get_aec: Late-bound accessor for the optional
            :class:`~getpatter.audio.aec.NlmsEchoCanceller`.
        get_audio_filter: Late-bound accessor for the optional
            :class:`~getpatter.providers.base.AudioFilter`
            (``agent.audio_filter`` — Krisp / DeepFilterNet).
        get_vad: Late-bound accessor for the active
            :class:`~getpatter.providers.base.VADProvider`
            (``agent.vad`` or the auto-loaded Silero instance).
        input_sample_rate: Sample rate (Hz) of the inbound PCM16 stream when
            ``input_is_mulaw_8k`` is ``False``. Default ``16000``. When it is
            NOT 16 kHz the chain resamples to 16 kHz through a per-call
            ``StatefulResampler`` instead of forwarding the bytes at the wrong
            rate (a PCM16 @ 8 kHz carrier previously reached STT pitched up an
            octave). Ignored for the mu-law path (always 8 kHz -> 16 kHz).
        high_pass_hz: Optional high-pass / DC-block cutoff (Hz). When set, a
            stateful :class:`~getpatter.audio.high_pass.BiquadHighPass` runs as
            the FIRST stage, on the decoded signal before the resampler.
            ``None`` (default) disables it.
        agc: Optional automatic gain control. ``None`` (default) disables it;
            an :class:`~getpatter.models.AgcConfig` enables a stateful
            :class:`~getpatter.audio.agc.Agc` after noise suppression and
            before VAD/STT.
    """

    def __init__(
        self,
        *,
        input_is_mulaw_8k: bool,
        get_aec: Callable[[], object | None],
        get_audio_filter: Callable[[], "AudioFilter | None"],
        get_vad: Callable[[], "VADProvider | None"],
        input_sample_rate: int = PIPELINE_SAMPLE_RATE,
        high_pass_hz: int | None = None,
        agc: object | None = None,
    ) -> None:
        if input_sample_rate <= 0:
            raise ValueError(f"input_sample_rate must be > 0; got {input_sample_rate}.")
        self._input_is_mulaw_8k = input_is_mulaw_8k
        self._input_sample_rate = int(input_sample_rate)
        self._get_aec = get_aec
        self._get_audio_filter = get_audio_filter
        self._get_vad = get_vad
        # Lazily created on the first mulaw frame (mirrors the handler's
        # historical lazy import) so PCM-native deployments never touch
        # audioop.
        self._resampler_8k_to_16k: "StatefulResampler | None" = None
        # Lazily created when a non-mulaw PCM stream arrives at a rate other
        # than 16 kHz (``input_sample_rate`` -> 16 kHz).
        self._pcm_resampler: "StatefulResampler | None" = None
        # Warn-once latches for the fail-open stage wrappers.
        self._filter_warned = False
        self._hpf_warned = False
        self._agc_warned = False

        # The high-pass runs on the decoded signal BEFORE the resampler, so its
        # operating rate is the carrier rate (8 kHz for mu-law, otherwise
        # ``input_sample_rate``). Built eagerly here; a bad cutoff disables the
        # stage (fail-open) rather than breaking call construction.
        self._hpf = None
        if high_pass_hz is not None:
            hpf_rate = 8000 if input_is_mulaw_8k else self._input_sample_rate
            try:
                from getpatter.audio.high_pass import BiquadHighPass

                self._hpf = BiquadHighPass(
                    cutoff_hz=float(high_pass_hz), sample_rate=hpf_rate
                )
            except Exception as exc:  # noqa: BLE001 - fail-open
                logger.warning(
                    "high_pass_hz=%s could not be initialised at %d Hz "
                    "(stage disabled): %s",
                    high_pass_hz,
                    hpf_rate,
                    exc,
                )

        # AGC runs at the pipeline rate (16 kHz) after noise suppression.
        self._agc = None
        if agc is not None:
            try:
                from getpatter.audio.agc import Agc

                self._agc = Agc(
                    sample_rate=PIPELINE_SAMPLE_RATE,
                    target_rms_dbfs=agc.target_rms_dbfs,
                    max_gain_db=agc.max_gain_db,
                    speech_floor_dbfs=agc.speech_floor_dbfs,
                    attack_ms=agc.attack_ms,
                    release_ms=agc.release_ms,
                    limiter_ceiling=agc.limiter_ceiling,
                )
            except Exception as exc:  # noqa: BLE001 - fail-open
                logger.warning("agc could not be initialised (stage disabled): %s", exc)

    async def process(self, audio_bytes: bytes) -> InputFrame:
        """Run one inbound media frame through decode -> AEC -> filter -> VAD.

        Never raises for filter failures (fail-open passthrough, warn once).
        VAD failures are swallowed per-frame at DEBUG — parity with the
        pre-extraction handler behaviour.
        """
        # ---- decode -> HPF -> resample ----------------------------------------
        # Decode mu-law (or take PCM as-is), then run the optional high-pass /
        # DC-block on the decoded signal at the CARRIER rate, then normalise to
        # 16 kHz. Running the HPF before the resampler keeps DC / rumble out of
        # the interpolation and matches the canonical APM order.
        if self._input_is_mulaw_8k:
            from getpatter.audio.transcoding import mulaw_to_pcm16

            pcm = mulaw_to_pcm16(audio_bytes)
            pcm = self._apply_high_pass(pcm)
            if self._resampler_8k_to_16k is None:
                from getpatter.audio.transcoding import create_resampler_8k_to_16k

                self._resampler_8k_to_16k = create_resampler_8k_to_16k()
            pcm = self._resampler_8k_to_16k.process(pcm)
        else:
            pcm = self._apply_high_pass(audio_bytes)
            if self._input_sample_rate != PIPELINE_SAMPLE_RATE:
                # PCM16 at a non-16 kHz rate: resample instead of forwarding the
                # bytes at the wrong rate (which pitches the caller up/down and
                # corrupts STT alignment).
                if self._pcm_resampler is None:
                    from getpatter.audio.transcoding import StatefulResampler

                    self._pcm_resampler = StatefulResampler(
                        src_rate=self._input_sample_rate,
                        dst_rate=PIPELINE_SAMPLE_RATE,
                    )
                pcm = self._pcm_resampler.process(pcm)

        # ---- AEC ---- subtract estimated TTS bleed before filter/VAD/STT.
        # Pass-through until the canceller has enough far-end history to
        # fill its filter window (~128 ms), then converges over the next
        # 0.5-2 s of TTS-only frames.
        aec = self._get_aec()
        if aec is not None:
            pcm = aec.process_near_end(pcm)  # type: ignore[attr-defined]

        # ---- audio_filter ---- noise suppression (Krisp / DeepFilterNet).
        # AFTER AEC, BEFORE VAD per the AudioFilter ABC contract. Fail-open:
        # a broken filter must never take down the call audio path.
        audio_filter = self._get_audio_filter()
        if audio_filter is not None:
            try:
                filtered = await audio_filter.process(pcm, PIPELINE_SAMPLE_RATE)
            except Exception as exc:
                self._warn_filter_once(audio_filter, exc)
            else:
                if isinstance(filtered, (bytes, bytearray)):
                    pcm = bytes(filtered)
                else:
                    self._warn_filter_once(
                        audio_filter,
                        TypeError(
                            f"process() returned {type(filtered).__name__}, expected bytes"
                        ),
                    )

        # ---- AGC ---- normalise level AFTER noise suppression, BEFORE VAD/STT.
        # Speech-selective + peak-limited; fail-open like audio_filter.
        if self._agc is not None:
            try:
                gained = self._agc.process(pcm)
            except Exception as exc:  # noqa: BLE001 - fail-open
                self._warn_agc_once(exc)
            else:
                if isinstance(gained, (bytes, bytearray)):
                    pcm = bytes(gained)
                else:  # pragma: no cover - defensive
                    self._warn_agc_once(
                        TypeError(
                            f"Agc.process returned {type(gained).__name__}, expected bytes"
                        )
                    )

        # ---- VAD ---- feed the (filtered) frame; at most one event back.
        vad = self._get_vad()
        vad_event: "Optional[VADEvent]" = None
        if vad is not None:
            try:
                vad_event = await vad.process_frame(pcm, PIPELINE_SAMPLE_RATE)
            except Exception as exc:  # pragma: no cover - defensive
                logger.debug("VAD process_frame failed: %s", exc)
                vad_event = None

        return InputFrame(pcm=pcm, vad_event=vad_event, vad_configured=vad is not None)

    def flush(self) -> None:
        """Flush and discard the inbound resampler tails (call teardown)."""
        if self._resampler_8k_to_16k is not None:
            self._resampler_8k_to_16k.flush()
            self._resampler_8k_to_16k = None
        if self._pcm_resampler is not None:
            self._pcm_resampler.flush()
            self._pcm_resampler = None

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _apply_high_pass(self, pcm: bytes) -> bytes:
        """Run the optional high-pass stage (fail-open passthrough, warn once)."""
        if self._hpf is None:
            return pcm
        try:
            out = self._hpf.process(pcm)
        except Exception as exc:  # noqa: BLE001 - fail-open
            if not self._hpf_warned:
                self._hpf_warned = True
                logger.warning(
                    "high-pass stage failed; passing audio through unfiltered "
                    "(further failures logged at DEBUG): %s",
                    exc,
                )
            else:
                logger.debug("high-pass stage failed (passthrough): %s", exc)
            return pcm
        return out if isinstance(out, (bytes, bytearray)) else pcm

    def _warn_agc_once(self, exc: Exception) -> None:
        """WARN on the first AGC failure, DEBUG afterwards (fail-open)."""
        if not self._agc_warned:
            self._agc_warned = True
            logger.warning(
                "AGC stage failed; passing audio through without gain "
                "(further failures logged at DEBUG): %s",
                exc,
            )
        else:
            logger.debug("AGC stage failed (passthrough): %s", exc)

    def _warn_filter_once(self, audio_filter: object, exc: Exception) -> None:
        """WARN on the first filter failure, DEBUG afterwards (fail-open)."""
        if not self._filter_warned:
            self._filter_warned = True
            logger.warning(
                "audio_filter %s failed; passing audio through unfiltered "
                "(further failures logged at DEBUG): %s",
                type(audio_filter).__name__,
                exc,
            )
        else:
            logger.debug(
                "audio_filter %s failed (passthrough): %s",
                type(audio_filter).__name__,
                exc,
            )

File: libraries/typescript/src/services/input-chain.ts
/**
 * Inbound audio processing chain for pipeline mode (slice 1 of the
 * StreamHandler decomposition — see ``docs/architecture/pipeline-stages.md``).
 *
 * Owns the stateless-to-STT half of ``handleAudio``:
 *
 *     decode (mulaw -> PCM16) -> high-pass/DC-block (opt) -> resample
 *     -> AEC near-end (opt) -> ``agent.audioFilter``/NS (opt)
 *     -> AGC (opt) -> VAD frame feed
 *
 * and returns the processed frame plus at most one VAD event per frame. The
 * handler keeps everything downstream for this slice (VAD-event handling,
 * self-hearing gate, inbound ring buffer, ``beforeSendToStt`` hook, STT feed)
 * so the change stays reviewable.
 *
 * Stage-order contract (fixed) — canonical WebRTC APM order
 * ``HPF -> AEC -> NS -> VAD -> AGC`` adapted for telephony (AGC pulled ahead
 * of VAD/STT so the recogniser sees a level-normalised signal):
 * - The high-pass / DC-block runs FIRST, on the decoded signal BEFORE the
 *   resampler, so DC / sub-cutoff rumble never reaches the echo canceller.
 * - AEC runs before the noise suppressor so the suppressor never disturbs the
 *   canceller's far-end/near-end alignment.
 * - ``audioFilter`` (noise suppression) runs AFTER AEC and BEFORE VAD, per the
 *   ``AudioFilter`` interface contract — the VAD then benefits from the
 *   cleaned signal.
 * - AGC runs AFTER noise suppression and BEFORE VAD/STT so it normalises the
 *   cleaned signal (never amplifying suppressed noise) and the level-normalised
 *   frame helps both the VAD energy gate and the STT recogniser.
 *
 * Every new stage (HPF, AGC) is OPT-IN, provider-agnostic, and fail-open
 * (a raising stage degrades to passthrough, warns once, then DEBUG).
 *
 * The filter wrapper is fail-open with a warn-once policy: a rejecting (or
 * non-Buffer-returning) filter degrades to passthrough of the pre-filter PCM,
 * logs one WARNING, keeps logging at DEBUG, and keeps being attempted — a
 * transient provider hiccup must not permanently strip noise suppression, and
 * a permanent one must never break the call audio path.
 *
 * AEC / audio-filter / VAD are resolved through late-bound getter callables
 * rather than captured at construction: ``StreamHandler`` populates ``aec`` /
 * ``autoVad`` during ``initPipeline`` (after this chain is constructed) and
 * the unit suites assign them directly on handler instances.
 *
 * Mirrors Python ``getpatter/services/input_chain.py``.
 */

import { getLogger } from '../logger';
import { mulawToPcm16, StatefulResampler } from '../audio/transcoding';
import { BiquadHighPass } from '../audio/high-pass';
import { Agc } from '../audio/agc';
import type { AudioFilter, VADEvent, VADProvider, AgcConfig } from '../types';

/** The pipeline's internal processing rate: inbound carrier audio is always
 *  normalised to PCM16 mono @ 16 kHz before AEC / filter / VAD / STT. */
export const PIPELINE_SAMPLE_RATE = 16000;

/** Minimal AEC surface the chain needs (avoids a hard import of audio/aec). */
export interface NearEndProcessor {
  processNearEnd(pcm: Buffer): Buffer;
}

/** Result of pushing one carrier media frame through the input chain. */
export interface ProcessedInputFrame {
  /** Decoded / resampled / AEC'd / filtered PCM16 16 kHz bytes — exactly
   *  what should reach the self-hearing gate and STT. */
  readonly pcm16k: Buffer;
  /** The VAD event emitted for this frame, if any. */
  readonly vadEvent: VADEvent | null;
  /** True when a VAD provider is configured for this call — the handler's
   *  self-hearing gate applies even while VAD is error-disabled. */
  readonly vadConfigured: boolean;
}

/** Constructor options for {@link InputProcessingChain}. */
export interface InputProcessingChainOptions {
  /** The per-call stateful 8k→16k resampler (injected so the handler can
   *  flush its tail on call close — see ``flushResamplers``). Used for the
   *  mu-law decode path. */
  readonly resampler: StatefulResampler;
  /** Late-bound accessor for the optional NLMS echo canceller. */
  readonly getAec: () => NearEndProcessor | null | undefined;
  /** Late-bound accessor for the optional ``agent.audioFilter``. */
  readonly getAudioFilter: () => AudioFilter | null | undefined;
  /** Late-bound accessor for the active VAD (``agent.vad`` ?? auto-Silero). */
  readonly getVad: () => VADProvider | null | undefined;
  /** H4 hot-path guard: max ms to wait for VAD inference before treating
   *  the frame as silent. Defaults to 25 ms (the pre-extraction value). */
  readonly vadTimeoutMs?: number;
  /** When ``true`` (default) inbound frames are G.711 mu-law @ 8 kHz and are
   *  decoded + upsampled to 16 kHz. When ``false`` they are linear PCM16 at
   *  {@link inputSampleRate}. All current carriers are mu-law. */
  readonly inputIsMulaw8k?: boolean;
  /** Sample rate (Hz) of the inbound PCM16 stream when ``inputIsMulaw8k`` is
   *  ``false``. Default 16000. When not 16 kHz the chain resamples to 16 kHz
   *  instead of forwarding the bytes at the wrong rate. */
  readonly inputSampleRate?: number;
  /** Optional high-pass / DC-block cutoff (Hz). Runs as the FIRST stage, on
   *  the decoded signal before the resampler. Undefined disables it. */
  readonly highPassHz?: number;
  /** Optional inbound AGC. ``true`` uses {@link AgcConfig} defaults, an object
   *  tunes it, ``false``/undefined disables it. Runs after noise suppression
   *  and before VAD/STT. */
  readonly agc?: boolean | AgcConfig;
}

/** HPF -> resample -> AEC -> audioFilter -> AGC -> VAD for inbound audio. */
export class InputProcessingChain {
  private readonly opts: InputProcessingChainOptions;
  /** Warn-once latch for the fail-open audio-filter wrapper. */
  private filterWarned = false;
  /** Warn-once latches for the fail-open HPF / AGC wrappers. */
  private hpfWarned = false;
  private agcWarned = false;
  /** Set after a VAD failure to suppress log spam for the rest of the call. */
  private vadDisabled = false;

  private readonly inputIsMulaw8k: boolean;
  private readonly inputSampleRate: number;
  /** Optional first-stage high-pass / DC-block (built once, reused). */
  private readonly hpf: BiquadHighPass | null = null;
  /** Optional AGC (built once, reused). */
  private readonly agc: Agc | null = null;
  /** Lazily-created resampler for the non-mulaw, non-16 kHz PCM path. */
  private pcmResampler: StatefulResampler | null = null;

  constructor(opts: InputProcessingChainOptions) {
    this.opts = opts;
    this.inputIsMulaw8k = opts.inputIsMulaw8k ?? true;
    this.inputSampleRate = opts.inputSampleRate ?? PIPELINE_SAMPLE_RATE;
    if (this.inputSampleRate <= 0) {
      throw new Error(`inputSampleRate must be > 0, got ${this.inputSampleRate}`);
    }

    // High-pass runs on the decoded signal BEFORE the resampler, so its
    // operating rate is the carrier rate (8 kHz for mu-law, else the PCM
    // input rate). A bad cutoff disables the stage (fail-open).
    if (opts.highPassHz !== undefined) {
      const hpfRate = this.inputIsMulaw8k ? 8000 : this.inputSampleRate;
      try {
        this.hpf = new BiquadHighPass({ cutoffHz: opts.highPassHz, sampleRate: hpfRate });
      } catch (err) {
        getLogger().warn(
          `highPassHz=${opts.highPassHz} could not be initialised at ${hpfRate} Hz ` +
            `(stage disabled): ${String(err)}`,
        );
      }
    }

    // AGC runs at the pipeline rate (16 kHz) after noise suppression.
    if (opts.agc) {
      const cfg: AgcConfig = opts.agc === true ? {} : opts.agc;
      try {
        this.agc = new Agc({
          sampleRate: PIPELINE_SAMPLE_RATE,
          targetRmsDbfs: cfg.targetRmsDbfs,
          maxGainDb: cfg.maxGainDb,
          speechFloorDbfs: cfg.speechFloorDbfs,
          attackMs: cfg.attackMs,
          releaseMs: cfg.releaseMs,
          limiterCeiling: cfg.limiterCeiling,
        });
      } catch (err) {
        getLogger().warn(`agc could not be initialised (stage disabled): ${String(err)}`);
      }
    }
  }

  /**
   * Run one inbound media frame through decode -> AEC -> filter -> VAD.
   *
   * Never rejects for filter failures (fail-open passthrough, warn once).
   * A VAD failure disables VAD for the rest of the call (warn once) —
   * parity with the pre-extraction handler behaviour.
   */
  async process(audioBuffer: Buffer): Promise<ProcessedInputFrame> {
    // Decode -> HPF -> resample. Run the optional high-pass / DC-block on the
    // decoded signal at the CARRIER rate (before the resampler), then
    // normalise to 16 kHz. Running the HPF first keeps DC / rumble out of the
    // interpolation and matches the canonical APM order.
    let pcm16k: Buffer;
    if (this.inputIsMulaw8k) {
      // All current carriers (Twilio, Plivo, Telnyx-PCMU) deliver mu-law 8 kHz.
      let pcm8k = mulawToPcm16(audioBuffer);
      pcm8k = this.applyHighPass(pcm8k);
      pcm16k = this.opts.resampler.process(pcm8k);
    } else {
      pcm16k = this.applyHighPass(audioBuffer);
      if (this.inputSampleRate !== PIPELINE_SAMPLE_RATE) {
        // PCM16 at a non-16 kHz rate: resample instead of forwarding the bytes
        // at the wrong rate (which pitches the caller and corrupts STT).
        if (!this.pcmResampler) {
          this.pcmResampler = new StatefulResampler({
            srcRate: this.inputSampleRate,
            dstRate: PIPELINE_SAMPLE_RATE,
          });
        }
        pcm16k = this.pcmResampler.process(pcm16k);
      }
    }

    // Acoustic echo cancellation — subtract estimated TTS bleed from the
    // mic stream before filter/VAD/STT see it. Pass-through until the
    // canceller has enough far-end history to fill its filter window
    // (~128 ms), then converges over the next 0.5–2 s of TTS-only frames.
    const aec = this.opts.getAec();
    if (aec) {
      pcm16k = aec.processNearEnd(pcm16k);
    }

    // Noise suppression (``agent.audioFilter`` — DeepFilterNet/Krisp).
    // AFTER AEC, BEFORE VAD per the AudioFilter contract. Fail-open: a
    // broken filter must never take down the call audio path.
    const audioFilter = this.opts.getAudioFilter();
    if (audioFilter) {
      try {
        const filtered = await audioFilter.process(pcm16k, PIPELINE_SAMPLE_RATE);
        if (Buffer.isBuffer(filtered)) {
          pcm16k = filtered;
        } else {
          this.warnFilterOnce(
            audioFilter,
            `process() returned ${typeof filtered}, expected Buffer`,
          );
        }
      } catch (err) {
        this.warnFilterOnce(audioFilter, String(err));
      }
    }

    // AGC — normalise level AFTER noise suppression, BEFORE VAD/STT.
    // Speech-selective + peak-limited; fail-open like audioFilter.
    if (this.agc) {
      try {
        const gained = this.agc.process(pcm16k);
        if (Buffer.isBuffer(gained)) {
          pcm16k = gained;
        } else {
          this.warnAgcOnce(`process() returned ${typeof gained}, expected Buffer`);
        }
      } catch (err) {
        this.warnAgcOnce(String(err));
      }
    }

    // External VAD (e.g. Silero) when configured — runs on the (AEC'd,
    // filtered, AGC'd) frame so noise suppression improves barge-in robustness.
    const activeVad = this.opts.getVad();
    let vadEvent: VADEvent | null = null;
    if (activeVad && !this.vadDisabled) {
      try {
        // H4: protect hot path against slow ONNX inference — if VAD takes
        // longer than ``vadTimeoutMs``, treat the frame as silent and continue.
        const vadPromise = activeVad.processFrame(pcm16k, PIPELINE_SAMPLE_RATE);
        let vadTimeoutId: ReturnType<typeof setTimeout>;
        const timeoutPromise = new Promise<null>((resolve) => {
          vadTimeoutId = setTimeout(() => resolve(null), this.opts.vadTimeoutMs ?? 25);
        });
        vadEvent = (await Promise.race([vadPromise, timeoutPromise])) ?? null;
        clearTimeout(vadTimeoutId!);
      } catch (err) {
        this.disableVad();
        getLogger().warn(
          `VAD processFrame failed — disabling VAD for this call: ${String(err)}`,
        );
      }
    }

    return { pcm16k, vadEvent, vadConfigured: Boolean(activeVad) };
  }

  /**
   * Disable VAD for the rest of the call. Also invoked by the handler when
   * VAD-event handling throws, preserving the pre-extraction semantics where
   * one try/catch covered both inference and event handling.
   */
  disableVad(): void {
    this.vadDisabled = true;
  }

  /** Whether VAD has been error-disabled for this call. */
  isVadDisabled(): boolean {
    return this.vadDisabled;
  }

  /** Run the optional high-pass stage (fail-open passthrough, warn once). */
  private applyHighPass(pcm: Buffer): Buffer {
    if (!this.hpf) return pcm;
    try {
      const out = this.hpf.process(pcm);
      return Buffer.isBuffer(out) ? out : pcm;
    } catch (err) {
      if (!this.hpfWarned) {
        this.hpfWarned = true;
        getLogger().warn(
          `high-pass stage failed; passing audio through unfiltered ` +
            `(further failures logged at DEBUG): ${String(err)}`,
        );
      } else {
        getLogger().debug(`high-pass stage failed (passthrough): ${String(err)}`);
      }
      return pcm;
    }
  }

  private warnAgcOnce(detail: string): void {
    if (!this.agcWarned) {
      this.agcWarned = true;
      getLogger().warn(
        `AGC stage failed; passing audio through without gain ` +
          `(further failures logged at DEBUG): ${detail}`,
      );
    } else {
      getLogger().debug(`AGC stage failed (passthrough): ${detail}`);
    }
  }

  private warnFilterOnce(audioFilter: AudioFilter, detail: string): void {
    const name = audioFilter.constructor?.name ?? 'AudioFilter';
    if (!this.filterWarned) {
      this.filterWarned = true;
      getLogger().warn(
        `audioFilter ${name} failed; passing audio through unfiltered ` +
          `(further failures logged at DEBUG): ${detail}`,
      );
    } else {
      getLogger().debug(`audioFilter ${name} failed (passthrough): ${detail}`);
    }
  }
}
